native/src/main/java/io/ballerina/stdlib/serdes/Deserializer.java

```diff
@@ -52,251 +66,335 @@
+
+    private static Object getUnionTypeValueFromMessage(DynamicMessage dynamicMessage, Type type) {
+
+        Descriptor messageDescriptor = dynamicMessage.getDescriptorForType();
+
+        for (var entry : dynamicMessage.getAllFields().entrySet()) {
+            Object value = entry.getValue();
+            FieldDescriptor fieldDescriptor = entry.getKey();
+
+            // Handle null value
+            if (fieldDescriptor.getName().equals(NULL_FIELD_NAME) && (boolean) value) {
+                return null;
+            }
+
+            // Handle array values
+            if (fieldDescriptor.isRepeated()) {
+                String fieldName = fieldDescriptor.getName();
+                String[] tokens = fieldName.split(TYPE_SEPARATOR);
+                String ballerinaTypeName = tokens[0];
+                int dimention = Integer.parseInt(tokens[1].split(SEPARATOR)[1]);
+                ArrayType arrayType = getBallerinaArrayTypeFromUnion((UnionType) type, ballerinaTypeName, dimention);
+                return getArrayTypeValueFromMessage(
+                        value,
+                        arrayType.getElementType(),
+                        messageDescriptor,
+                        dimention,
+                        ballerinaTypeName);
+            } else if (value instanceof ByteString && fieldDescriptor.getName().contains(ARRAY_FIELD_NAME)) {
+                // Handle byte array values
+                ByteString byteString = (ByteString) value;
+                return ValueCreator.createArrayValue(byteString.toByteArray());
+            } else if (value instanceof DynamicMessage
+                    && !fieldDescriptor.getMessageType().getName().contains(DECIMAL_VALUE)) {
+                // Handle record values
+                String fieldName = fieldDescriptor.getName();
+                String[] tokens = fieldName.split(TYPE_SEPARATOR);
+                String ballerinaType = tokens[0];
+                RecordType recordType = getBallerinaRecordTypeFromUnion((UnionType) type, ballerinaType);
+                return getRecordTypeValueFromMessage((DynamicMessage) value, recordType);
+            } else {
+                // Handle primitive values
+                return getPrimitiveTypeValueFromMessage(value);
+            }
+        }
+
+        throw createSerdesError(UNSUPPORTED_DATA_TYPE + type.getName(), SERDES_ERROR);
+    }
+
+    private static Object getArrayTypeValueFromMessage(
+            Object value,
+            Type elementType,
+            Descriptor messageDescriptor) {
+        return getArrayTypeValueFromMessage(value, elementType, messageDescriptor, -1, null);
+    }
+
+    private static Object getArrayTypeValueFromMessage(
+            Object value,
+            Type elementType,
+            Descriptor messageDescriptor,
+            int dimensions,
+            String ballerinaTypeNamePrefixOfUnionMemberOrRecordField) {
```

Comment:
This formatting too. Might be an IDE issue. Let's not split lines for each parameter.

Reply:
Updated accordingly